src/lib.rs

```diff
@@ -447,15 +447,28 @@ impl Options {
         line
     }
 
-    /// Derive a usage message from a set of options.
+
+    /// Derive a formatted message from a set of options.
     pub fn usage(&self, brief: &str) -> String {
+        self.usage_with_format(|opts|
+            format!("{}\n\nOptions:\n{}\n", brief, opts.collect::<Vec<String>>().join("\n")))
+    }
+
+    /// Derive a custom formatted message from a set of options. The formatted options provided to
+    /// a closure as an iterator.
+    pub fn usage_with_format<'a, F: FnMut(Box<Iterator<Item=String> + 'a>) -> String>(&'a self, mut formatter: F) -> String {
```

Comment:
Could you make this generic over any Iterator, instead of hardcoding it to a trait object?

Reply:
@Kimundi, without `impl Trait` specialization that's in the RFC - I didn't think it was possible to make it return a generic iterator. Perhaps, I missed something. Could you please explain?

Reply:
After a short discussion on IRC, it seems I was wrong about this being able to be made generic.

However, would it make sense to have the closure take &mut Iterator? just to not force a box allocation on other callers of this interface.

Reply:
Done :)